YakimS/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,231 @@
+    {
+        if(keySet.contains(key))
+        {
+            throw new KeyAlreadyExistsException();
+        }
+        else {
+            keySet.add(key);
+        }
+
+        Node nodeToAdd = new Node(key, value);
+        if(root == null)
+        {
+            root = nodeToAdd;
+            return;
+        }
+
+        /*
+        *   Moving down the tree, while choosing the path down (through right or left son)
+        *  randomly, till the current node is null, and then assign to the added node
+        *  to that null current node
+        */
+        Node currentNode = root;
+        Random random = new Random();
+        while(currentNode != null)
+        {
+            if (random.nextBoolean())
+            {
+                if (currentNode.getLeftSon() == null)
+                {
+                    currentNode.setLeftSon(nodeToAdd);
+                    return;
+                }
+                else
+                {
+                    currentNode = currentNode.getLeftSon();
+                }
+            }
+            else
+            {
+                if (currentNode.getRightSon() == null)
+                {
+                    currentNode.setRightSon(nodeToAdd);
+                    return;
+                }
+                else
+                {
+                    currentNode = currentNode.getRightSon();
+                }
+            }
+        }
+    }
+
+    /**
+     * Prints all the ancestors key of the given key.
+     * If there is'nt such key in the tree, nothing will be printed.
+     */
+    public void printAncestors(int key)
+    {
+        printAncestors(key,root);
+    }
+    private boolean printAncestors(int key, Node root)
+    {
+        if(root == null)
+        {
+            return false;
+        }
+        else if(root.getKey() == key)
+        {
+            return true;
+        }
+        else if(printAncestors(key,root.getLeftSon()) ||
+                printAncestors(key,root.getRightSon()))
+        {
+            System.out.print( root==this.root ? root.getKey() : root.getKey() + ", ");
+            return true;
+        }
+        else {
+            return false;
+        }
+    }
+
+    /**
+     * This function find the lowest common ancestor by finding the path to
+     * each given key, than return the last common number in both paths.
+     *
+     * @param key1 - first key to find common ancestor
+     * @param key2 - second key to find common ancestor
+     * @return lowest common ancestor of the two given keys. If the keys are equal than
+     *          the function will return the key number.
+     * @throws KeyNotFoundException if one or both of the keys are not in the tree
+     */
+    public int getLowestCommonAncestor(int key1, int key2) throws KeyNotFoundException {
+
+        // create paths to the two keys given
+        List<Integer> firstKeyPath = getKeyPath(key1, root, new LinkedList<>());

```

Comment:
The assignment explicitly asked to avoid storing nodes in additional data structures. While technically you are storing keys instead of nodes, in practice it is the same. Could you change the code to avoid that?